### ISSUE TITLE
Span lost in tracegen

### ISSUE THREAD

**Issue by judyQD** [jaegertracing/jaeger#2928]
**Describe the bug**
I am using tracegen to do the performance testing. 
When I set the duration with 600s, it generated spans: 14831883. But the flushed span is only 473153.
I also enlarge the max_queue_size, and short the flush interval. But the gap is still huge

Below is the log of tracegen:
[root@pl-038-control-01 tragegen]# kubectl logs tracegen-6d77458c88-4f4xj -f
2021-04-07T07:12:04.927Z        INFO    command-line-arguments/main.go:62       Initialized global tracer
2021-04-07T07:12:04.928Z        INFO    tracegen/config.go:55   Duration has been set, traces will be forced to 0
2021-04-07T07:22:04.928Z        INFO    **tracegen/worker.go:103  Worker 0 generated 14831883 traces      {"worker": 0}**
2021-04-07T07:22:04.929Z        INFO    command-line-arguments/main.go:66       Waiting 1.5sec for metrics to flush
2021-04-07T07:22:06.429Z        DEBUG   zap/logger.go:45        closing tracer
2021-04-07T07:22:06.429Z        DEBUG   zap/logger.go:45        closing reporter
2021-04-07T07:22:06.429Z        DEBUG   zap/logger.go:45        **flushed span num: 473153**

The cmd to start tracegen:
./tracegen-linux-amd64 -duration 600s

The configuration of tracegen is as below:
        env:
          - name: JAEGER_ENDPOINT
            value: http://10.99.1.89:14268/api/traces
          **- name: JAEGER_REPORTER_MAX_QUEUE_SIZE
            value: "20000"**
          - name: JAEGER_REPORTER_FLUSH_INTERVAL
            value: 5ms
          - name: JAEGER_SERVICE_NAME
            value: replica1_600_1

**To Reproduce**
1. 
2.
3.

**Expected behavior**
I think the number of span generated should be equal to the number of span flushed to collector.

**Screenshots**

**Version (please complete the following information):**
 - OS: [e.g. Linux]
 - Jaeger version: [e.g. 1.8]
 - Deployment: [e.g.  Docker, ...]

**What troubleshooting steps did you try?**
Try to follow https://www.jaegertracing.io/docs/latest/troubleshooting/ and describe how far you were able to progress and/or which steps did not work.
I add a global var to counter the flushed span.  jaeger-client-go/reporter.go: 
func (r *remoteReporter) processQueue() {
        // flush causes the Sender to flush its accumulated spans and clear the buffer
        flush := func() {
                if flushed, err := r.sender.Flush(); err != nil {
                        r.metrics.ReporterFailure.Inc(int64(flushed))
                        r.logger.Error(fmt.Sprintf("failed to flush Jaeger spans to server: %s", err.Error()))
                } else if flushed > 0 {
                        r.metrics.ReporterSuccess.Inc(int64(flushed))
                }
        }

        timer := time.NewTicker(r.bufferFlushInterval)
        for {
                select {
                case <-timer.C:
                        flush()
                case item := <-r.queue:
                        atomic.AddInt64(&r.queueLength, -1)
                        switch item.itemType {
                        case reporterQueueItemSpan:
                                span := item.span
                                if flushed, err := r.sender.Append(span); err != nil {
                                        r.metrics.ReporterFailure.Inc(int64(flushed))
                                        r.logger.Error(fmt.Sprintf("error reporting Jaeger span %q: %s", span.OperationName(), err.Error()))
                                } else if flushed > 0 {
                                
                                        r.metrics.ReporterSuccess.Inc(int64(flushed))
                                        **flush_count = flush_count + int64(flushed)**
                                        
                                        // to reduce the number of gauge stats, we only emit queue length on flush
                                        r.metrics.ReporterQueueLength.Update(atomic.LoadInt64(&r.queueLength))
                                        **//r.logger.Debugf("flushed %d spans", flushed)**
                                }
                                span.Release()

**Additional context**

**Comment by jpkrohling** [comment]
How many spans are seen by the collector? Take a look at the /metrics endpoint in the admin port (14269 for all-in-one) and report what you see as part of the metric `jaeger_spans_received_total`

**Comment by judyQD** [comment]
> 
> 
> How many spans are seen by the collector? Take a look at the /metrics endpoint in the admin port (14269 for all-in-one) and report what you see as part of the metric `jaeger_spans_received_total`

I did the test again, result is shown as below:
jaeger_collector_spans_received_total{debug="false",format="jaeger",svc="replica1_600_1",transport="http"} 526593

Tracegen log:
2021-04-08T02:44:47.274Z        INFO    command-line-arguments/main.go:62       Initialized global tracer
2021-04-08T02:44:47.280Z        INFO    tracegen/config.go:55   Duration has been set, traces will be forced to 0
2021-04-08T02:54:47.286Z        INFO    tracegen/worker.go:103  Worker 0 generated 13117705 traces      {"worker": 0}
2021-04-08T02:54:47.287Z        INFO    command-line-arguments/main.go:66       Waiting 1.5sec for metrics to flush
2021-04-08T02:54:48.789Z        DEBUG   zap/logger.go:45        closing tracer
2021-04-08T02:54:48.789Z        DEBUG   zap/logger.go:45        closing reporter
2021-04-08T02:54:48.789Z        DEBUG   zap/logger.go:45        judy flushed span num: 497174

Total generated span: 13117705, flushed span is 497174, while collector received span is 526593. This is realy wired.

**Comment by yurishkuro** [comment]
You may be generating spans faster than the SDK is able to send them out. The SDK has similar metrics for the number of spans created, finished, reported, dropped (although I am not sure tracegen enables those metrics).

Also, if you're reporting spans over UDP and the collector is not able to process them, some UDP packets will get lost.

**Comment by judyQD** [comment]
> 
> 
> You may be generating spans faster than the SDK is able to send them out. The SDK has similar metrics for the number of spans created, finished, reported, dropped (although I am not sure tracegen enables those metrics).
> 
> Also, if you're reporting spans over UDP and the collector is not able to process them, some UDP packets will get lost.

Thanks for your info. All the spans are reported over HTTP. 
The main issue I am talking about the span is lost inside of tracegen. The generated spans are much more than the span reported to Collector.

**Comment by jpkrohling** [comment]
Checking the metrics on the collector side would confirm that your counter is accurate. As @yurishkuro mentioned, the Jaeger Client for Go goes drop spans if more data is generated than it can flush, which would explain what you are seeing. Unfortunately, tracegen isn't exposing the client metrics, so that we can't confirm this is the case. I just opened a PR that exposes those metrics using the Prometheus client.

**Comment by judyQD** [comment]
@jpkrohling @yurishkuro Thanks for your help. Yes, you are right, Jaeger Client drops spans. 
Could you explain more, after creating the span, where are the spans stored in? Is there any ways to decrease the drop number? Thx.

**Comment by jpkrohling** [comment]
Spans are placed in an internal queue and picked up by workers. If the queue gets full, spans get dropped, which is what typically happens when your workers are too slow. This seems to be your case, as the network seems to be the bottleneck.

The typical way of sending spans from clients to Jaeger is via UDP to the agent, so that items are processed quickly in the client. The agent will batch the spans and send them via gRPC to the collector, which is usually more efficient than what the client can do. If you absolutely have to use HTTP for the communication between the client (tracegen) and the collector, try tweaking your client configuration, like by changing the `JAEGER_REPORTER_MAX_QUEUE_SIZE` env var. 

In any case, I think the network will still be the bottleneck...

**Comment by judyQD** [comment]
> 
> 
> Spans are placed in an internal queue and picked up by workers. If the queue gets full, spans get dropped, which is what typically happens when your workers are too slow. This seems to be your case, as the network seems to be the bottleneck.
> 
> The typical way of sending spans from clients to Jaeger is via UDP to the agent, so that items are processed quickly in the client. The agent will batch the spans and send them via gRPC to the collector, which is usually more efficient than what the client can do. If you absolutely have to use HTTP for the communication between the client (tracegen) and the collector, try tweaking your client configuration, like by changing the `JAEGER_REPORTER_MAX_QUEUE_SIZE` env var.
> 
> In any case, I think the network will still be the bottleneck...

You are right I am using tracegen send spans to collector directly via HTTP transport. 
1. could you point out the internal queue stores spans. I want to enlarge the queue and have a try.
2. cfg JAEGER_REPORTER_MAX_QUEUE_SIZE has been set to 30000, still many spans are dropped
3. I also set the flush interval: JAEGER_REPORTER_FLUSH_INTERVAL to 5ms, still many spans are dropped

**Comment by jpkrohling** [comment]
This would be the relevant reporter:

https://github.com/jaegertracing/jaeger-client-go/blob/99ac6dc786f933252951f55baf6b0d73db17c4dc/reporter.go#L194-L205
